### PR TITLE
Add amalgamation traits for NFT CollectionId and ItemId

### PR DESCRIPTION
**Pull request by jasl**
I don't know this is the right direction but I believe this would help when people interacting with NFT traits return values in their own pallets

Basically I'm imitating how [fungible traits](https://github.com/paritytech/substrate/blob/master/frame/support/src/traits/tokens/fungible.rs#L38) does

Background:
I'm learning NFTs pallet so I'm building my own pallet that invoking pallet_nfts via nonfungibles_v2 interface,
I tried:
- Create collections in my pallet and deposit an event
- Mint items in my pallet

then, I found few problems:
- `create_collection` returns `collection_id` but can't be used in event's payload because ```the trait `Clone` is not implemented for `<<T as pallet::Config>::NonFungibles as frame_support::traits::tokens::nonfungibles_v2::Inspect<<T as frame_system::Config>::AccountId>>::CollectionId```
- `mint_into` requires `ItemConfig` but it can be constructed out of pallet_nfts

and a question:

How to construct CollectionId and ItemId from a 3rd pallet?

### CODE REVIEW

**Comment by jasl** [comment]
Add a non topic change (perhaps I need update the title?)

**Comment by jasl** [comment]
@jsidorenko Could you help to take a look? (don't know how to assign reviewers...)

**Comment by jsidorenko** [comment]
Hey @jasl ,
> How to construct CollectionId and ItemId from a 3rd pallet?

Could you pls provide some code snippet or push your code to GH and put a link here?
I don't quite understand the problem you're facing when constructing the ItemId, won't the `1234u32.into()` do the magic?

**Comment by jasl** [comment]
> Hey @jasl ,
> 
> > How to construct CollectionId and ItemId from a 3rd pallet?
> 
> 
> 
> Could you pls provide some code snippet or push your code to GH and put a link here?
> 
> I don't quite understand the problem you're facing when constructing the ItemId, won't the `1234u32.into()` do the magic?

I remeber I tried but not work,
I'll re-check and make a public repo for you if not works

**Comment by jasl** [comment]
> Hey @jasl ,
> 
> > How to construct CollectionId and ItemId from a 3rd pallet?
> 
> Could you pls provide some code snippet or push your code to GH and put a link here? I don't quite understand the problem you're facing when constructing the ItemId, won't the `1234u32.into()` do the magic?

I pushed reproducible code

version 1: https://github.com/jasl/learn_nft/blob/main/src/lib.rs
my pallet extending pallet_nft

```
error[E0603]: type alias `NextCollectionId` is private
   --> src/lib.rs:116:18
    |
116 |                 pallet_nfts::NextCollectionId::<T, I>::get().unwrap_or(CollectionIdOf::<T, I>::initial_value());
    |                              ^^^^^^^^^^^^^^^^ private type alias
    |
note: the type alias `NextCollectionId` is defined here
   --> /Users/jasl/.cargo/git/checkouts/substrate-7e08433d4c370a21/2cb4825/frame/nfts/src/lib.rs:59:9
    |
59  | pub use pallet::*;
    |         ^^^^^^^^^

error[E0277]: the trait bound `<T as pallet_nfts::Config<I>>::ItemId: From<u32>` is not satisfied
   --> src/lib.rs:147:10
    |
147 |                 0u32.into(),
    |                      ^^^^ the trait `From<u32>` is not implemented for `<T as pallet_nfts::Config<I>>::ItemId`
    |
    = note: required for `u32` to implement `Into<<T as pallet_nfts::Config<I>>::ItemId>`
help: consider further restricting the associated type
    |
135 |         ) -> DispatchResult where <T as pallet_nfts::Config<I>>::ItemId: From<u32> {
    |                             ++++++++++++++++++++++++++++++++++++++++++++++++++++++

Some errors have detailed explanations: E0277, E0603.
For more information about an error, try `rustc --explain E0277`.
```

version 2: https://github.com/jasl/learn_nft/blob/another-case/src/lib.rs
use nonfungibles_v2 trait 

```
error[E0277]: the trait bound `<<T as pallet::Config>::NFTCollection as frame_support::traits::tokens::nonfungibles_v2::Inspect<<T as frame_system::Config>::AccountId>>::CollectionId: Clone` is not satisfied
  --> src/lib.rs:81:45
   |
81 |         CollectionCreated { worker: T::AccountId, collection_id: CollectionIdOf<T> },
   |                                                   ^^^^^^^^^^^^^ the trait `Clone` is not implemented for `<<T as pallet::Config>::NFTCollection as frame_support::traits::tokens::nonfungibles_v2::Inspect<<T as frame_system::Config>::AccountId>>::CollectionId`

error[E0369]: binary operation `==` cannot be applied to type `&<<T as pallet::Config>::NFTCollection as frame_support::traits::tokens::nonfungibles_v2::Inspect<<T as frame_system::Config>::AccountId>>::CollectionId`
  --> src/lib.rs:81:45
   |
81 |         CollectionCreated { worker: T::AccountId, collection_id: CollectionIdOf<T> },
   |                                                   ^^^^^^^^^^^^^

error[E0277]: `<<T as pallet::Config>::NFTCollection as frame_support::traits::tokens::nonfungibles_v2::Inspect<<T as frame_system::Config>::AccountId>>::CollectionId` doesn't implement `std::fmt::Debug`
  --> src/lib.rs:81:45
   |
81 |         CollectionCreated { worker: T::AccountId, collection_id: CollectionIdOf<T> },
   |                                                   ^^^^^^^^^^^^^ `<<T as pallet::Config>::NFTCollection as frame_support::traits::tokens::nonfungibles_v2::Inspect<<T as frame_system::Config>::AccountId>>::CollectionId` cannot be formatted using `{:?}` because it doesn't implement `std::fmt::Debug`
   |
   = help: the trait `std::fmt::Debug` is not implemented for `<<T as pallet::Config>::NFTCollection as frame_support::traits::tokens::nonfungibles_v2::Inspect<<T as frame_system::Config>::AccountId>>::CollectionId`
   = note: required for `&<<T as pallet::Config>::NFTCollection as frame_support::traits::tokens::nonfungibles_v2::Inspect<<T as frame_system::Config>::AccountId>>::CollectionId` to implement `std::fmt::Debug`
   = note: required for the cast from `&<<T as pallet::Config>::NFTCollection as frame_support::traits::tokens::nonfungibles_v2::Inspect<<T as frame_system::Config>::AccountId>>::CollectionId` to the object type `dyn std::fmt::Debug`

Some errors have detailed explanations: E0277, E0369.
For more information about an error, try `rustc --explain E0277`.
error: could not compile `pallet-nft_computing` due to 3 previous errors
warning: build failed, waiting for other jobs to finish...
```

**Comment by jsidorenko** [comment]
Hi @jasl, I'm wondering why don't you want to connect the nfts pallet in a loosely coupled way?
Then you will be able to access the mint_into and create_collection methods. In the case of the create_collection(), you won't need to pass the collection id and deal with the NextCollectionId storage item.
You can find the example of connecting the pallet in this PR: https://github.com/paritytech/substrate/pull/12565

**Comment by jsidorenko** [comment]
Sorry, I just checked the second example where you used the traits

**Comment by jsidorenko** [comment]
so, the item_id problem I was able able to solve locally by adding ` + From<u32>` to type's definition in your pallet

**Comment by jasl** [comment]
> Hi @jasl, I'm wondering why don't you want to connect the nfts pallet in a loosely coupled way? Then you will be able to access the mint_into and create_collection methods. In the case of the create_collection(), you won't need to pass the collection id and deal with the NextCollectionId storage item. You can find the example of connecting the pallet in this PR: #12565

My another case (https://github.com/jasl/learn_nft/blob/another-case/src/lib.rs) use that way, I'll learn from that PR first.

thank you

**Comment by jsidorenko** [comment]
@jasl regarding the Copy/Clone issue, I found it doesn't complain if you provide the ItemId/CollectionId via the runtime config: https://github.com/paritytech/substrate/pull/12565/files#diff-d757e0d131cebbebb0176e5448e53b54970ebb64f1c5d7f01537a4e221e587b0R175-R176

File: frame/support/src/traits/tokens/nonfungible_v2.rs [hunk 1]
```diff
@@ -24,6 +24,7 @@
 //! For an NFT API that has dual-level namespacing, the traits in `nonfungibles` are better to
 //! use.
 
+use super::misc::ItemId;
```
Comment (jsidorenko) [inline frame/support/src/traits/tokens/nonfungible_v2.rs:27]: I would add the `ItemId, CollectionId and DestroyWitness` into the `pub use misc::{` of the `frame/support/src/traits/tokens.rs` and import later via `use frame_support::traits::tokens::{CollectionId, ItemId}`

Comment (jasl) [reply frame/support/src/traits/tokens/nonfungible_v2.rs:27]: Can you help? I think this PR can be closed later

**Comment by jasl** [comment]
> @jasl regarding the Copy/Clone issue, I found it doesn't complain if you provide the ItemId/CollectionId via the runtime config: https://github.com/paritytech/substrate/pull/12565/files#diff-d757e0d131cebbebb0176e5448e53b54970ebb64f1c5d7f01537a4e221e587b0R175-R176

I'll try this tonight, thank you for metion this.

File: frame/support/src/traits/tokens/nonfungible_v2.rs (hunk 1 above)
Comment (jsidorenko) [reply frame/support/src/traits/tokens/nonfungible_v2.rs:27]: Just open the `frame/support/src/traits/tokens.rs` file, find `pub use misc::{` and add there `ItemId, CollectionId, DestroyWitness`

Comment (jasl) [reply frame/support/src/traits/tokens/nonfungible_v2.rs]: Done

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2503232

File: frame/support/src/traits/tokens/nonfungibles_v2.rs [hunk 2]
```diff
@@ -27,19 +27,22 @@
...
+};
 use codec::{Decode, Encode};
 use sp_runtime::TokenError;
 use sp_std::prelude::*;
 
 /// Trait for providing an interface to many read-only NFT-like sets of items.
 pub trait Inspect<AccountId> {
 	/// Type for identifying an item.
-	type ItemId;
+	type ItemId: ItemId;
```
Comment (bkchr) [inline frame/support/src/traits/tokens/nonfungibles_v2.rs:41]:
```suggestion
	type ItemId: Parameter;
```

File: frame/support/src/traits/tokens/nonfungibles_v2.rs [hunk 3]
```diff
@@ -27,19 +27,22 @@
...
 
 	/// Type for identifying a collection (an identifier for an independent collection of
 	/// items).
-	type CollectionId;
+	type CollectionId: CollectionId;
```
Comment (bkchr) [inline frame/support/src/traits/tokens/nonfungibles_v2.rs:45]:
```suggestion
	type CollectionId: Parameter;
```

File: frame/support/src/traits/tokens/nonfungibles_v2.rs [hunk 4]
```diff
@@ -193,7 +196,7 @@ pub trait Create<AccountId, CollectionConfig>: Inspect<AccountId> {
 /// Trait for providing the ability to destroy collections of nonfungible items.
 pub trait Destroy<AccountId>: Inspect<AccountId> {
 	/// The witness data needed to destroy an item.
-	type DestroyWitness;
+	type DestroyWitness: DestroyWitness;
```
Comment (bkchr) [inline frame/support/src/traits/tokens/nonfungibles_v2.rs:199]:
```suggestion
	type DestroyWitness: Parameter;
```

File: frame/support/src/traits/tokens/nonfungibles_v2.rs (hunk 2 above)
Comment (bkchr) [reply frame/support/src/traits/tokens/nonfungibles_v2.rs:41]: Maybe also `+ Copy`, but not sure we should have this here.

File: frame/support/src/traits/tokens/nonfungibles_v2.rs [hunk 5]
```diff
@@ -27,19 +27,22 @@
 //! Implementations of these traits may be converted to implementations of corresponding
 //! `nonfungible` traits by using the `nonfungible::ItemOf` type adapter.
 
-use crate::dispatch::{DispatchError, DispatchResult};
+use crate::{
+	dispatch::{DispatchError, DispatchResult},
```
Comment (bkchr) [inline frame/support/src/traits/tokens/nonfungibles_v2.rs:31]:
```suggestion
	dispatch::{DispatchError, DispatchResult, Parameter},
```

File: frame/support/src/traits/tokens/nonfungibles_v2.rs [hunk 6]
```diff
@@ -27,19 +27,22 @@
...
+	traits::tokens::{CollectionId, DestroyWitness, ItemId},
```
Comment (bkchr) [inline frame/support/src/traits/tokens/nonfungibles_v2.rs:32]:
```suggestion
```

File: frame/support/src/traits/tokens/nonfungible_v2.rs [hunk 7]
```diff
@@ -25,15 +25,18 @@
 //! use.
 
 use super::nonfungibles_v2 as nonfungibles;
-use crate::{dispatch::DispatchResult, traits::Get};
+use crate::{
+	dispatch::DispatchResult,
```
Comment (bkchr) [inline frame/support/src/traits/tokens/nonfungible_v2.rs:29]:
```suggestion
	dispatch::{DispatchResult, Parameter},
```

File: frame/support/src/traits/tokens/nonfungible_v2.rs [hunk 8]
```diff
@@ -25,15 +25,18 @@
...
+};
 use codec::{Decode, Encode};
 use sp_runtime::TokenError;
 use sp_std::prelude::*;
 
 /// Trait for providing an interface to a read-only NFT-like item.
 pub trait Inspect<AccountId> {
 	/// Type for identifying an item.
-	type ItemId;
+	type ItemId: ItemId;
```
Comment (bkchr) [inline frame/support/src/traits/tokens/nonfungible_v2.rs:39]:
```suggestion
	type ItemId: Parameter;
```

File: frame/support/src/traits/tokens/nonfungible_v2.rs [hunk 9]
```diff
@@ -25,15 +25,18 @@
 //! use.
 
 use super::nonfungibles_v2 as nonfungibles;
-use crate::{dispatch::DispatchResult, traits::Get};
+use crate::{
+	dispatch::DispatchResult,
+	traits::{tokens::ItemId, Get},
```
Comment (bkchr) [inline frame/support/src/traits/tokens/nonfungible_v2.rs:30]:
```suggestion
	traits::Get,
```

File: frame/support/src/traits/tokens/misc.rs [hunk 10]
```diff
@@ -210,6 +210,36 @@ pub trait BalanceConversion<InBalance, AssetId, OutBalance> {
...
+/// Simple amalgamation trait to collect together properties for a DestroyWitness under one roof.
+pub trait DestroyWitness:
+	FullCodec + Copy + Eq + PartialEq + Debug + scale_info::TypeInfo + MaxEncodedLen
+{
+}
+impl<T: FullCodec + Copy + Eq + PartialEq + Debug + scale_info::TypeInfo + MaxEncodedLen>
+	DestroyWitness for T
+{
+}
+
```
Comment (bkchr) [inline frame/support/src/traits/tokens/misc.rs:242]:
```suggestion
```

**Review by bkchr — APPROVED** [review]
We don't need new traits for all of them, we can just use what is already there.

File: frame/support/src/traits/tokens/nonfungibles_v2.rs (hunk 2 above)
Comment (jasl) [reply frame/support/src/traits/tokens/nonfungibles_v2.rs]: @jsidorenko what do you think?

**Comment by jasl** [comment]
I'll apply all Basti's suggests then squash them

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-check-each-crate
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2504249

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2504238

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/substrate/-/jobs/2504240

**Comment by jsidorenko** [comment]
can you pls run `cargo +nightly fmt`?

**Review by jsidorenko — APPROVED** [review]

**Comment by bkchr** [comment]
> @bkchr I basically follow how AssetId does, do you think AssetId (and maybe Balance) can apply this?

Good questions :P IDK, I would need to check the code.

**Comment by bkchr** [comment]
bot fmt

**Comment by jsidorenko** [comment]
P. S. @jasl I think we should simplify the collection's creation via traits and remove the config param.
If the advanced (non-default) settings are needed, they should be set via separate methods.
I've opened a PR for this today: https://github.com/paritytech/substrate/pull/13572

**Comment by jasl** [comment]
> #13572

GREAT! Thank you !
After this one merge, I'll refactor my code

**Comment by jasl** [comment]
@bkchr Could you help to merge this?

**Comment by jsidorenko** [comment]
bot merge

**Pull request merged**